### PR TITLE
Capture the SNS topic ARN under the 'messaging.destination.name' span attribute.

### PR DESCRIPTION
**Pull request by tduncan**
Addresses #10030

### CODE REVIEW

**Review by zeitlinger — APPROVED** [review]

**Comment by laurit** [comment]
Should the same change also be made for aws 1 instrumentation? Ideally I think we should change the SNS spans to follow messaging semantic conventions like we did for SQS, though this is probably too much effort and out of scope for this issue.

**Comment by tduncan** [comment]
@laurit, I'd be happy to update the AWS SDK v1 instrumentation as well though I don't see explicit support for SNS as I do in the v2 instrumentation.

**Comment by laurit** [comment]
@tduncan I think in aws1 sdk you'd need to add an accessor in https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/main/instrumentation/aws-sdk/aws-sdk-1.11/library/src/main/java/io/opentelemetry/instrumentation/awssdk/v1_11/RequestAccess.java that lets you reflectively retrieve the value and create an attribute extractor similar to https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/main/instrumentation/aws-sdk/aws-sdk-1.11/library/src/main/java/io/opentelemetry/instrumentation/awssdk/v1_11/AwsSdkExperimentalAttributesExtractor.java Can't use `AwsSdkExperimentalAttributesExtractor` because experimental (non spec) attributes are collected only when enabled with a flag. And add that extractor to https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/2199cc5b13d097bf73609fa96244c30701e61ffd/instrumentation/aws-sdk/aws-sdk-1.11/library/src/main/java/io/opentelemetry/instrumentation/awssdk/v1_11/AwsSdkInstrumenterFactory.java#L44 and https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/2199cc5b13d097bf73609fa96244c30701e61ffd/instrumentation/aws-sdk/aws-sdk-1.11/library/src/main/java/io/opentelemetry/instrumentation/awssdk/v1_11/AwsSdkInstrumenterFactory.java#L48 Hopefully that would be enough (besides aws1 tests also run camel-2.20 tests, these also rely on the aws1 instrumentation)

File: instrumentation/aws-sdk/aws-sdk-2.2/testing/src/main/groovy/io/opentelemetry/instrumentation/awssdk/v2_2/AbstractAws2ClientTest.groovy [hunk 1]
```diff
@@ -156,6 +160,16 @@ abstract class AbstractAws2ClientTest extends AbstractAws2ClientCoreTest {
     "S3"      | "CreateBucket"      | "PUT"  | "UNKNOWN"                              | s3ClientBuilder()       | { c -> c.createBucket(CreateBucketRequest.builder().bucket("somebucket").build()) }              | ""
     "S3"      | "GetObject"         | "GET"  | "UNKNOWN"                              | s3ClientBuilder()       | { c -> c.getObject(GetObjectRequest.builder().bucket("somebucket").key("somekey").build()) }     | ""
     "Kinesis" | "DeleteStream"      | "POST" | "UNKNOWN"                              | KinesisClient.builder() | { c -> c.deleteStream(DeleteStreamRequest.builder().streamName("somestream").build()) }          | ""
+    "Sns"     | "Publish"           | "POST" | "d74b8436-ae13-5ab4-a9ff-ce54dfea72a0" | SnsClient.builder()     | { c -> c.publish(PublishRequest.builder().message("somemessage").topicArn("somearn").build()) }  | """
```
Comment (laurit) [inline instrumentation/aws-sdk/aws-sdk-2.2/testing/src/main/groovy/io/opentelemetry/instrumentation/awssdk/v2_2/AbstractAws2ClientTest.groovy:163]: besides `topicArn` there is also `targetArn` should that also be handled?

Comment (tduncan) [reply instrumentation/aws-sdk/aws-sdk-2.2/testing/src/main/groovy/io/opentelemetry/instrumentation/awssdk/v2_2/AbstractAws2ClientTest.groovy:163]: I think so. Do you have any thoughts on which should be preferred in the event that both were provided (and different!)?

**Comment by tduncan** [comment]
@laurit, thanks for the tips! I'll explore the v1 SDK instrumentation in more detail.

File: instrumentation/aws-sdk/aws-sdk-2.2/testing/src/main/groovy/io/opentelemetry/instrumentation/awssdk/v2_2/AbstractAws2ClientTest.groovy (hunk 1 above)
Comment (laurit) [reply instrumentation/aws-sdk/aws-sdk-2.2/testing/src/main/groovy/io/opentelemetry/instrumentation/awssdk/v2_2/AbstractAws2ClientTest.groovy:163]: I think it depends on what aws does.

Comment (tduncan) [reply instrumentation/aws-sdk/aws-sdk-2.2/testing/src/main/groovy/io/opentelemetry/instrumentation/awssdk/v2_2/AbstractAws2ClientTest.groovy:163]: I tested what happens when both `TopicArn` and `TargetArn` are provided and SNS responds with an error:
`Invalid parameter: TopicArn Reason: Both TopicArn and TargetArn specified. Use only one or the other.`

File: instrumentation/aws-sdk/aws-sdk-1.11/library/src/main/java/io/opentelemetry/instrumentation/awssdk/v1_11/SnsAttributesExtractor.java [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+import io.opentelemetry.context.Context;
+import io.opentelemetry.instrumentation.api.instrumenter.AttributesExtractor;
+import io.opentelemetry.semconv.SemanticAttributes;
+import javax.annotation.Nullable;
+
+public class SnsAttributesExtractor implements AttributesExtractor<Request<?>, Response<?>> {
+  @Override
+  public void onStart(AttributesBuilder attributes, Context parentContext, Request<?> request) {
+    String destination = findMessageDestination(request.getOriginalRequest());
+    if (destination != null) {
```
Comment (laurit) [inline instrumentation/aws-sdk/aws-sdk-1.11/library/src/main/java/io/opentelemetry/instrumentation/awssdk/v1_11/SnsAttributesExtractor.java:21]: Usually we use https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/1ee51e410e90de9a54e1f840c63b222d390f921f/instrumentation-api/src/main/java/io/opentelemetry/instrumentation/api/internal/AttributesExtractorUtil.java#L22-L27 which does the null check

File: instrumentation/aws-sdk/aws-sdk-2.2/library/src/main/java/io/opentelemetry/instrumentation/awssdk/v2_2/AwsSdkRequestType.java [hunk 3]
```diff
@@ -15,7 +16,10 @@ enum AwsSdkRequestType {
   S3(request("aws.bucket.name", "Bucket")),
   SQS(request("aws.queue.url", "QueueUrl"), request("aws.queue.name", "QueueName")),
   KINESIS(request("aws.stream.name", "StreamName")),
-  DYNAMODB(request("aws.table.name", "TableName"));
+  DYNAMODB(request("aws.table.name", "TableName")),
+  SNS(
+      request(SemanticAttributes.MESSAGING_DESTINATION_NAME.getKey(), "TargetArn"),
```
Comment (laurit) [inline instrumentation/aws-sdk/aws-sdk-2.2/library/src/main/java/io/opentelemetry/instrumentation/awssdk/v2_2/AwsSdkRequestType.java:21]: perhaps add a comment here that only one of these can be set

**Review by laurit — APPROVED** [review]

**Review by kenfinnigan — APPROVED** [review]

**Pull request merged**